### ISSUE TITLE
GC Finalization Cycle Warnings

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#2728]
Building in debug mode gives a lot of GC warnings when building documentation of Macaulay2Doc. Is this normal? Is it worth investigating when it started?

When building the first time, here is when the first one shows up on my machine:
```m2
 -- capturing example results for "Set - Set"                               
GC Warning: Finalization cycle involving 0x7f9bc3f50a00
```
The address seems to be always the same, but I'm not sure how to debug that.

**Comment by DanGrayson** [comment]
Finalization cycles lead to memory leaks (uncollected garbage), so it's well worth it to eliminate them.

To figure it out, one would have to run M2 under gdb, and maybe set a break point to stop the first time the gc finalization routine is called with that pointer.  Then one could at least see what type of thing it is.

**Comment by mahrud** [comment]
How do you do that without knowing what that pointer is for?

**Comment by DanGrayson** [comment]
When the finalization routine in libgc is called, and the breakpoint triggers a stop, ascend the stack until you see what sort of thing the pointer is.

**Comment by mahrud** [comment]
GC_finalize is called thousands of times, how do you only break when it is finalizing that address?

**Comment by mahrud** [comment]
Also, unless I'm misunderstanding, isn't GC_finalize usually called long after a pointer is not used by any function in the stack anymore, hence being collected?

**Comment by mikestillman** [comment]
Yes, I think that it might be difficult to analyze that (find the pointer).  I'll try to do so as well.

**Comment by DanGrayson** [comment]
> GC_finalize is called thousands of times, how do you only break when it is finalizing that address?
```
'break ... if COND'
     Set a breakpoint with condition COND; evaluate the expression COND
     each time the breakpoint is reached, and stop only if the value is
     nonzero--that is, if COND evaluates as true.  '...' stands for one
     of the possible arguments described above (or no argument)
     specifying where to break.  *Note Break Conditions: Conditions, for
     more information on breakpoint conditions.
```

**Comment by DanGrayson** [comment]
> Also, unless I'm misunderstanding, isn't GC_finalize usually called long after a pointer is not used by any function in the stack anymore, hence being collected? How would I know what sort of thing the pointer is if nothing in the stack uses it? (In other words, if a routine in the stack is still using the pointer, why would it be collected?!)

The breakpoint would be set in the routine used to register a finalizer, not in the routine that does the finalization.
